### PR TITLE
Added dynamic dashboard layouts

### PR DESCRIPTION
**Pull request by lasanthaS**
1. Added dynamic dashboard layouts
2. Modified the existing layouts to be compatible with the dynamic layouts

### CODE REVIEW

File: apps/portal/js/designer.js [hunk 1]
```diff
@@ -1872,12 +2025,61 @@ $(function () {
                 renderComponentToolbar(findComponent(id));
             });
             listenLayout();
+            
+            var gridsterContainer = $('.gridster > ul');
+            
+            var w = Math.ceil($('.gridster').width() / 12) - 10;
```
Comment (manuranga) [inline apps/portal/js/designer.js]: better to have a more meaningful name than w, eg minBlockWidth

File: apps/portal/js/designer.js [hunk 2]
```diff
@@ -1839,6 +1929,69 @@ $(function () {
...
+        return { "blocks": blocks };
+    }
+
+    /**
+     * update the layout after modification
+     */
+    var updateLayout = function() {
+        
+        // extract the layout from the designer (gridster) and save it
+        var jsonLayout = convertToJsonLayout($('.gridster > ul'));
```
Comment (manuranga) [inline apps/portal/js/designer.js]: see if you can use gridster serialize instead.

File: apps/portal/js/designer.js [hunk 3]
```diff
@@ -1839,6 +1929,69 @@ $(function () {
...
+        var id;
+        
+        // find the current page index
+        for(var i = 0; i < ues.global.dashboard.pages.length; i++) {
+            if (ues.global.dashboard.pages[i].id === page.id) {
+                id = i;
+            }
+        }
+        
+        if (typeof id === 'undefined')
```
Comment (manuranga) [inline apps/portal/js/designer.js]: better to have { even for one lines

File: apps/portal/theme/templates/designer.jag [hunk 4]
```diff
@@ -532,5 +578,45 @@
     <button type="submit">Submit</button>
 </form><!-- /#ues-dashboard-upload-banner-form -->
 
+<script id="ues-component-box-list-hbs" type="text/x-handlebars-template">
+    
+    <ul>
+        {{#each blocks}}
```
Comment (manuranga) [inline apps/portal/theme/templates/designer.jag]: remove the duplicate of this in the below script tag.

File: apps/portal/js/ues-dashboards.js [hunk 5]
```diff
@@ -104,17 +103,58 @@
...
+    /**
+     * convert JSON layout to Bootstrap
+     * @param json
+     * @returns {*|jQuery}
+     */
+    var convertToBootstrapLayout = function(json, isFluid) {
+        var container = $('<div />').addClass(isFluid ? 'container-fluid' : 'container');
+        
+        JSONToBootstrap.convert(json.blocks, container, function(err) {
+            alert(err);
```
Comment (manuranga) [inline apps/portal/js/ues-dashboards.js]: use console error

File: apps/portal/css/common.css [hunk 6]
```diff
@@ -25,7 +25,15 @@
     text-align: right;
 }
 
-.ues-fullview-visible{
+.ues-fullview-visible { /* for the designer only */
+    position: absolute !important;
```
Comment (TanyaM) [inline apps/portal/css/common.css]: Check the possibility of getting rid of !important

File: apps/portal/js/dashboard.js [hunk 7]
```diff
@@ -13,30 +13,48 @@ $(function () {
...
+            
+            var uesComponent = $(this).closest('.ues-component'), 
+                component = findComponent(uesComponent.attr('id')), 
+                componentContainer = $(this).closest('.ues-component-box'), 
+                htmlBody = $('body');
+            
+            if (component.fullViewPoped) {
+                
+                // render normal view
+
```
Comment (TanyaM) [inline apps/portal/js/dashboard.js]: Remove new line

File: apps/portal/css/common.css (hunk 6 above)
Comment (lasanthaS) [inline apps/portal/css/common.css]: Cannot remove the !important tag as this class needs to override some of the styles added by 3rd party libraries.

File: apps/portal/js/dashboard.js (hunk 7 above)
Comment (lasanthaS) [inline apps/portal/js/dashboard.js]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

**Pull request merged**